### ISSUE TITLE
Contact Sensor Notifications not always working as expected.

### ISSUE THREAD

**Issue by madtempest** [0x5e/homebridge-tuya-platform#203]
- [X] I've read the [Readme - FAQ](https://github.com/0x5e/homebridge-tuya-platform#faq) and [Readme - Troubleshooting](https://github.com/0x5e/homebridge-tuya-platform#troubleshooting) section.

**Describe the bug**
Some contact sensors "mcs", when set in HomeKit to do notifications for activity: Time Any, People Off, do not create iOS notifications. Checking in the Smart Life app, I can see in the message center that the status of the sensors from Open to Closed is being recorded correctly (however based on reading this could be just the local MQTT).

One (1) of my contact sensors did successfully alert me around 30 minutes after the event. This is the only one of about 7 that is working correctly.

All contact sensors show in the Smart Life app as Bluetooth with a strikethrough including the sensor that worked with a 30 minute delay. All contact sensors show as "offline" in the Tuya IoT Developer platform. All sensors have a Smart Hub within 5 meters of their placement.

**Expected behavior**
Contact sensors, when set in HomeKit to do activity notifications, should create notifications when sensors open and close.

**Screenshots**
![IMG_7270](https://user-images.githubusercontent.com/14276889/211245061-1289e903-881c-4208-a31c-e202b2f2b502.jpg)

Smart Hubs are online, highlighted the "bluetooth strikethrough"

**Device info**
If the issue is related to a device, please provide the device info list and debug logs.

Will supply these as required.

**Notes**
The non-working sensors were previously configured using https://github.com/Kakise/homebridge-tuya-door-cloud until I saw that this plugin should do the same thing and as such this plugin is disabled in HomeBridge. Could be relevant?

**Comment by 0x5e** [comment]
I have the contact sensor and it's working on me.
As this plugin is cloud based, first you need to exclude the connetivity issue. disable ble and wifi on your phone, using celluar data only, and open the device status page on tuya app. when you close/open the door, the status should displayed on app in real time (within 3 seconds). If this is real time, the plugin should get working too.

**Comment by madtempest** [comment]
I suspected this was me only somehow.

Bluetooth off.
WiFi Off.
I have also changed the status of the display from "Door" back to "Contact Sensor" just in case for testing.

![IMG_7271](https://user-images.githubusercontent.com/14276889/211256529-73966b10-7c49-4734-8f31-4ca379dd73e0.jpg)

From IoT Console:

Device Information:
Product Name: Contact Sensor
Device ID: bf3f5561472d17b08dgtcv
Product Category: mcs
Device Status: Offline

![contact2](https://user-images.githubusercontent.com/14276889/211256983-efd0b1f0-b2ae-4f65-bf05-45b1d5c67e42.PNG)

Cloud debug device log shows the logs.

There is nothing in the HomeBridge Debug Log for the time period (14:50 -> 15:06) in the screenshot for the device ID bf3f5561472d17b08dgtcv which suggests that somehow I've borked something or the plugin can't query this device? Is there a way to test further please?

It is using standard instruction set.

**Comment by 0x5e** [comment]
try clean accessory cache, and confirm device notification service is subscribed on the website, and reboot homebridge, there's all I can thought.

**Comment by madtempest** [comment]
I'll try it and see what happens. Worse comes to worse I might reset them all. Thanks.

**Comment by madtempest** [comment]
So I have done some more testing and it appears to be related to session tokens perhaps?

Alerts are triggering - but only if the state changes and HomeBridge is restarted which renews or reauthenticates the session.

I'm using EU datacenter if that matters.

**Comment by madtempest** [comment]
I have reset the Smart Hubs (wfcon) and then reset the sensors and still the same behaviour where it only detects a change in the state on a restart of Homebridge. I even deleted the data when resetting the sensors and have deleted them as accessories from the cache after re-adding them.

Is there a way to get this plugin to debug and log only so I can track the MQ activity? I have too many devices and additional plugins to run debug mode for a long time - within 2 minutes debug logging was 30k lines long. I've converted the plugin to a child bridge but I can't get it to run in debug mode as a child bridge.

How often should the event 

[37m[1/23/2023, 10:34:22 AM] [39m[36m[TuyaPlatform][39m [90m[TuyaOpenMQ] Connecting to: ssl://m1.tuyaeu.com:8883[39m
[37m[1/23/2023, 10:34:23 AM] [39m[36m[TuyaPlatform][39m [90m[TuyaOpenAPI] Response:[39m

Fire? Because this is clearly the event that is checking the MQTT logs for things. Strangely, a newly added garage door opener seems to work fine...

**Comment by 0x5e** [comment]
You can use `grep` to filter the logs like this:
```
cat /path/to/your/logfile | grep TuyaOpenMQ
```
`TuyaOpenMQ` for the mqtt logs and `TuyaOpenAPI` for the api logs.

I think which data center doesn't matters, just confirm there's no error/warn logs, the network is okay, and you are not running multiple homebridge/homeassistant instance with one appkey/secret.

**Comment by madtempest** [comment]
At this stage, I'm definitely not running any other homebridge or homeassistants or other plugins using the same app key and secret.

Next step is 

- Delete the plugin from homebridge, also delete the persist JSON and any othere references to Tuya apps I can find in the docker container.
- Delete all of the items from Smart Life and re-add them one by one.

I'll let you know how it goes.

Thanks for your help and this awesome plugin.

**Comment by madtempest** [comment]
So I did the entire reset.

Still doesn't work.

The only thing I can think of next is this weirdness:

![image](https://user-images.githubusercontent.com/14276889/215252264-ffe4ab37-e844-424b-9fc2-d8fb998af471.png)

I don't know how or why there is two of them, or how I managed to somehow go over 1m API calls in less than a month when all I have used is this plugin and the Smart Life app.

Should I just start a new project and trash this one to fix the subscription issues?

**Comment by 0x5e** [comment]
Yes just have a try. Creating projcet or subscribe is free, just go ahead.

**Comment by madtempest** [comment]
So an update (Ive been busy).

Last weekend I created a new project, disassociated the Smart Life app from the old project, added it to the new one, added the keys etc - still doesn't work.

If I am not mistaken, is the logic path for notifications as follows:

Device performs action.
Device looks on the local network for the Smart Life app to deliver an MQTT message to.
If the Smart Life app is available, the device uses MQTT to tell the app the action was taken.
Smart Life app relays the message to the cloud project.
If the app is not available, the device uses its regular communication path to deliver the message to the cloud project.

Is that about right? I'm trying to ascertain where the failure is in my environment and clearly my environment alone.

Thanks.

**Comment by randymassey** [comment]
Just a thought from another user… I’m not sure if you have addressed this in a prior message, I see this is a fairly long message chain & I have not reviewed back on the entire thread. When I first set everything up, I tried to use the smartlife app & it did not work. I did, however, very quickly have success Setting everything up and then linking the Tuya app.You might’ve done that already, if so disregard… But just wanted to share a somewhat similar experience that I encountered & what ended up working for me.RandySent from my iPhoneOn Feb 17, 2023, at 9:20 PM, madtempest ***@***.***> wrote:﻿
So an update (Ive been busy).
Last weekend I created a new project, disassociated the Smart Life app from the old project, added it to the new one, added the keys etc - still doesn't work.
If I am not mistaken, is the logic path for notifications as follows:
Device performs action.
Device looks on the local network for the Smart Life app to deliver an MQTT message to.
If the Smart Life app is available, the device uses MQTT to tell the app the action was taken.
Smart Life app relays the message to the cloud project.
If the app is not available, the device uses its regular communication path to deliver the message to the cloud project.
Is that about right? I'm trying to ascertain where the failure is in my environment and clearly my environment alone.
Thanks.

**Comment by 0x5e** [comment]
@madtempest MQTT is a protocol used in cloud relay.
when device performs action, it will always send event to tuya cloud (if available) and local network tuya app (if any), and tuya cloud will transfer the event to the user's app or homebridge or homeassistant plugin who own this device. these two long connection are independent and message will be sent on both of them.

for sensors, they are mostly ble or zigbee (without wifi), their message will send to the gateway first, then gateway pass the message as above.

have you tried change device control mode? maybe cloud filter the non-standard message then plugin can't receive any updates. I see the old conversation that you can see device events on IoT Console (I assume it's almost realtime), so "sensor -> gateway -> tuya cloud" seems working.

**Comment by madtempest** [comment]
I know that the BLE gateways hubs (there's 4 to account for radio interference due to a lot of brick and concrete) show up as non-standard devices and as unsupported using the app. I did set them in the IoT console to be filtered out and not be read or anything but that didn't seem to work. I might have another go at this to filter out the hubs from the IoT console and subsequently the plugin.

Secondly, sorry for being dumb, but is the device control made changing the device instruction set from standard to the other one? I had considered that, I'll give that a shot next if removing the hubs doesn't make any difference.

Thank you very much for your assistance so far.

**Comment by madtempest** [comment]
> Just a thought from another user… I’m not sure if you have addressed this in a prior message, I see this is a fairly long message chain & I have not reviewed back on the entire thread. When I first set everything up, I tried to use the smartlife app & it did not work. I did, however, very quickly have success Setting everything up and then linking the Tuya app.You might’ve done that already, if so disregard… But just wanted to share a somewhat similar experience that I encountered & what ended up working for me.RandySent from my iPhoneOn Feb 17, 2023, at 9:20 PM, madtempest ***@***.***> wrote:﻿ So an update (Ive been busy). Last weekend I created a new project, disassociated the Smart Life app from the old project, added it to the new one, added the keys etc - still doesn't work. If I am not mistaken, is the logic path for notifications as follows: Device performs action. Device looks on the local network for the Smart Life app to deliver an MQTT message to. If the Smart Life app is available, the device uses MQTT to tell the app the action was taken. Smart Life app relays the message to the cloud project. If the app is not available, the device uses its regular communication path to deliver the message to the cloud project. Is that about right? I'm trying to ascertain where the failure is in my environment and clearly my environment alone. Thanks. —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

Thanks Randy - Smart Life was the app that worked originally which is why I haven't moved away from it and the app's been kept up to date but if all else otherwise fails, I'll give this a shot too.

It was working... and then the trial expired 6 months ago (and I was having issues anyway so I didn't notice) and it's been busted since.

**Comment by madtempest** [comment]
Okay so I'm starting to think the issue is either my Smart Life account potentially.

Here is the log from a whole Homebridge set up just for this plugin.
[homebridge.log.zip](https://github.com/0x5e/homebridge-tuya-platform/files/10830543/homebridge.log.zip)

When to test this, deleted my config for the old homebridge, uncached the accessories, deleted them out of Home app, and then added the new Homebridge and added the accessories.

What I noticed when testing is, when using the Smart Life app, the UI wouldn't update the status of the object if the app wasnt running as the object changed status. The log would show the change in status, ie MQTT was working I suspect, but the app wasn't.

This almost mirrors the fact that in my case, the plugin only changes its status when Homebridge is restarted.

So I'm going to create a new Smart Life/Tuya app account and see what happens next.

Appreciate your help so far.

**Comment by madtempest** [comment]
Finally got around to doing this:

New account still has the same UI glitch so I haven't bothered to try and add it to Homebridge.

Looking at how to hard reset the contact sensors as I am not trusting that the current method of using the Smart Life app is working.

**Comment by 0x5e** [comment]
@madtempest it's a long conversation and I'm a little forget about the previous context... have you tried changing device control mode on tuya website? guess this might work for you.

**Comment by madtempest** [comment]
Hi @0x5e sorry, thats the one thing I haven't tried. I wanted to clarify though:

"Secondly, sorry for being dumb, but is the device control made changing the device instruction set from standard to the other one? I had considered that, I'll give that a shot next if removing the hubs doesn't make any difference."

I'm gonna go with yes that is what you mean and try this today.

**Comment by 0x5e** [comment]
Yes, changing to "DP instruction" may show up more schema code. change it on website, then restart homebridge to take effect.

**Comment by madtempest** [comment]
So I never got around to doing the DP stuff, but I did finally find time to stuff around and do some clean up in my home network and some other stuff.

Good news, the state of my garage doors is now working but contact sensors are still a no-go.

I am wondering if this is related:

export default class Contac**e**Sensor extends BaseAccessory {

  requiredSchema() {
    return [SCHEMA_CODE.CONTACT_STATE];
  }

  configureServices() {
    const schema = this.getSchema(...SCHEMA_CODE.CONTACT_STATE);
    if (!schema) {
      return;
    }
    const service = this.accessory.getService(this.Service.ContactSensor)
    || this.accessory.addService(this.Service.ContactSensor);

    const { CONTACT_NOT_DETECTED, CONTACT_DETECTED } = this.Characteristic.ContactSensorState;
    service.getCharacteristic(this.Characteristic.ContactSensorState)
      .onGet(() => {
        const status = this.getStatus(schema.code)!;
        return status.value ? CONTACT_NOT_DETECTED : CONTACT_DETECTED;
      });
  }

}

Could this spelling mistake in the current Beta branch be why the service isn't working as all later spellings are correct. My JS is horrible, so I apologise if this makes no difference at all.

Ive also been looking at the raw logs in the iot.tuya.com website for the API explorer and noticed that the value returned for doorcontact_state is 

    {
      "code": "doorcontact_state",
      "value": false
    },
    
When the door is closed and true when the door is open.

Using the example of the Garage Door (which works state-wise for me), I can see that the code to implement the status of the contact is quite different. I'm guessing that the code for the sensors is more about efficiency?

Sorry for the very many questions but I finally have time to look at this again and some of my accessories are actually working that weren't. Its just constant sensors now.

**Comment by chewbanger** [comment]
> Yes, changing to "DP instruction" may show up more schema code. change it on website, then restart homebridge to take effect.

I had similar issues with my Zigbee Contact Sensors and this fixed the issue for me! Thanks
